### PR TITLE
Deprecate mediatedDevicesTypes in favor of mediatedDeviceTypes

### PR DESCRIPTION
**Pull request by tiraboschi**
**What this PR does / why we need it**:
With github.com/kubevirt/kubevirt/pull/8525
mediatedDevicesTypes got deprecated in favor of mediatedDeviceTypes on kubevirt/kubevirt.
Let's do the same on the HCO side for consistency.

The mutating webhook will take care of the migration to the new API and the validating webhook will
prevent inconsistencies between the two APIs.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2054863

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-16422
```

**Release note**:

```release-note
CR option mediatedDevicesTypes is deprecated in favor of mediatedDeviceTypes
```

### CODE REVIEW

File: controllers/operands/kubevirt_test.go [hunk 1]
```diff
@@ -586,30 +586,61 @@ Version: 1.2.3`)
 
 				mdevConf := foundResource.Spec.Configuration.MediatedDevicesConfiguration
 				Expect(mdevConf).ToNot(BeNil())
-				Expect(mdevConf.MediatedDevicesTypes).To(HaveLen(2))
-				Expect(mdevConf.MediatedDevicesTypes).To(ContainElements("nvidia-222", "nvidia-230"))
+				Expect(mdevConf.MediatedDeviceTypes).To(HaveLen(2))
+				Expect(mdevConf.MediatedDeviceTypes).To(ContainElements("nvidia-222", "nvidia-230"))
```
Comment (nunnatsa) [inline controllers/operands/kubevirt_test.go:590]: Do you ant also to check that `mdevConf.MediatedDevicesTypes` is empty?

File: controllers/operands/kubevirt_test.go [hunk 2]
```diff
@@ -633,25 +664,127 @@ Version: 1.2.3`)
...
-				Expect(mdevConf.NodeMediatedDeviceTypes[0].MediatedDevicesTypes).To(ContainElements("nvidia-223"))
+				Expect(mdevConf.NodeMediatedDeviceTypes[0].MediatedDeviceTypes).To(ContainElements("nvidia-223"))
 				Expect(mdevConf.NodeMediatedDeviceTypes[0].NodeSelector).To(HaveKeyWithValue("testLabel1", "true"))
-				Expect(mdevConf.NodeMediatedDeviceTypes[1].MediatedDevicesTypes).To(ContainElements("nvidia-229"))
+				Expect(mdevConf.NodeMediatedDeviceTypes[1].MediatedDeviceTypes).To(ContainElements("nvidia-229"))
 				Expect(mdevConf.NodeMediatedDeviceTypes[1].NodeSelector).To(HaveKeyWithValue("testLabel2", "true"))
 
 			})
-			It("should update the permitted host devices configuration from the HC", func() {
+			It("should propagate the mediated devices configuration from the HC with node selectors with new APIs", func() {
```
Comment (nunnatsa) [inline controllers/operands/kubevirt_test.go:676]: Not sure about "new API". This is very clear today. But one day (tomorrow?) it won't be new anymore...

File: pkg/webhooks/mutator/hyperConvergedMutator.go [hunk 3]
```diff
@@ -110,6 +110,25 @@ func (hcm *HyperConvergedMutator) mutateHyperConverged(_ context.Context, req ad
...
+		}
+		for i, hcoNodeMdevTypeConf := range hc.Spec.MediatedDevicesConfiguration.NodeMediatedDeviceTypes {
+			if len(hcoNodeMdevTypeConf.MediatedDevicesTypes) > 0 && len(hcoNodeMdevTypeConf.MediatedDeviceTypes) == 0 { //nolint SA1019
+				patches = append(patches, jsonpatch.JsonPatchOperation{
+					Operation: "add",
+					Path:      fmt.Sprintf("/spec/mediatedDevicesConfiguration/nodeMediatedDeviceTypes/%d/mediatedDeviceTypes", i),
+					Value:     hcoNodeMdevTypeConf.MediatedDevicesTypes, //nolint SA1019
+				})
+			}
+		}
```
Comment (nunnatsa) [inline pkg/webhooks/mutator/hyperConvergedMutator.go:129]: Do we also want to clear to MediatedDevicesTypes fields?

File: controllers/operands/kubevirt_test.go (hunk 2 above)
Comment (tiraboschi) [reply controllers/operands/kubevirt_test.go:676]: yes. it makes sense: let's call them mediatedDeviceTypes (explicitly) and deprecated (for better visibility).

File: pkg/webhooks/mutator/hyperConvergedMutator.go (hunk 3 above)
Comment (tiraboschi) [reply pkg/webhooks/mutator/hyperConvergedMutator.go:129]: No, I think we should not: they are deprecated but still there and some external code could also try to read it and we don't want to break it.

**Review by nunnatsa — APPROVED** [review]
/approve

**Pull request merged**